src/file.ts

```diff
@@ -2319,51 +2374,110 @@ class File extends ServiceObject<File> {
+      throw new Error(
+          `Max allowed expiration is seven days (${SEVEN_DAYS} seconds).`);
+    }
 
-            const parsedHost =
-                url.parse(config.cname || STORAGE_DOWNLOAD_BASE_URL);
-            const signedUrl = url.format({
-              protocol: parsedHost.protocol,
-              hostname: parsedHost.hostname,
-              pathname: config.cname ? name : this.bucket.name + '/' + name,
-              query,
-            });
+    const extensionHeaders = Object.assign({}, config.extensionHeaders);
+    extensionHeaders['host'] = 'storage.googleapis.com';
+    if (config.action === 'POST') {
+      extensionHeaders['x-goog-resumable'] = 'start';
+    }
+    if (config.contentMd5) {
+      extensionHeaders['content-md5'] = config.contentMd5;
+    }
+    if (config.contentType) {
+      extensionHeaders['content-type'] = config.contentType;
+    }
+
+    let signedHeaders = new Array<string>();  // should have host at least
+    signedHeaders = Object.keys(extensionHeaders);
+    signedHeaders = signedHeaders.map(header => header.toLowerCase());
+    signedHeaders.sort();
+
+    const extensionHeadersString = this.getCanonicalHeaders(extensionHeaders);
+
+    const datestamp = dateToISOString(now).split('T')[0];
+    const credentialScope = `${datestamp}/auto/storage/goog4_request`;
+
+    let queryParams = {} as SignedUrlQuery;
+
+    return this.storage.authClient.getCredentials()
+        .then((credentials) => {
+          const credential = `${credentials.client_email}/${credentialScope}`;
+
+          queryParams = {
+            'X-Goog-Algorithm': 'GOOG4-RSA-SHA256',
+            'X-Goog-Credential': credential,
+            'X-Goog-Date': dateToISOString(now),
+            'X-Goog-Expires': expiresPeriodInSeconds.toString(),
+            'X-Goog-SignedHeaders': signedHeaders.join(';'),
+          };
 
-            callback!(null, signedUrl);
+          const canonicalQueryParams = [];
+          const keys = Object.keys(queryParams);
+          for (const param of keys) {
+            keys.sort();
+            canonicalQueryParams.push(
+                param + '=' + encodeURIComponent(queryParams[param]));
+          }
+
+          const canonicalRequest = [
+            config.action,
+            config.resource,
+            canonicalQueryParams.join('&'),
+            extensionHeadersString,
+            signedHeaders.join(';'),
+            'UNSIGNED-PAYLOAD',
+          ].join('\n');
+
+          const canonicalRequestHash = crypto.createHash('sha256')
+                                           .update(canonicalRequest)
+                                           .digest('hex');
+
+          const blobToSign = [
+            'GOOG4-RSA-SHA256',
+            dateToISOString(now),
+            credentialScope,
+            canonicalRequestHash,
+          ].join('\n');
+
+          return this.storage.authClient.sign(blobToSign).catch((err) => {
+            throw new SigningError(err.message);
```

Comment:
fixed

Reply:
This is a good point. How should we go about doing this if we want it to throw (rejects with) `SigningError` while preserving. I couldn't find an existing pattern that I can look to. @bcoe @JustinBeckwith

Reply:
That's a really great question 😆  Curious on @ofrobots thoughts here as well.

Reply:
I'm not aware of the motivation behind using a custom error class; so I will refrain from commenting on that.

1. Can you copy over the `err.stack` property?
1. Is it possible for `authClient.sign` to throw the right 'type' of exception in the first place?

Reply:
(I was refactoring existing code so I'm just speculating here) I believe a `SigningError` is used to differentiate errors that arise because of some kind of missing/problems with signing, e.g. no email in credentials, rather than some issue with the code (which would be an `Error`). 

I think I'll go with 2. here changing the `sign` method, although I'm still curious if there's a way to correctly do this. Copying over `err.stack` feels too hacky for my taste.

Reply:
@jkwlui @ofrobots I've definitely used this approach before, bubbling a custom error type, as a way of offering a more contextual handling of certain types of errors; as an example in yargs, it allows us to print known error types as pretty text to the user, while still bubbling application level errors.

I believe that instead of copying `err.stack` there's a method available for subclasses of Error: `captureStackTrace`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error

Reply:
@bcoe the idea is to _not_ capture the stack trace, because we want the context to remain the same as the original.

@jkwlui I don't necessarily consider copying over the stack to be 'hacky'.